### PR TITLE
Publish battery voltage messages

### LINKED ISSUE #12: Publish power status messages
Finish figuring out relationship between ADC readings and currents and put into the CPP

### PR DESCRIPTION
**Pull request by jim-v**
Closes #12

### CODE REVIEW

File: src/motor_hardware.cc [hunk 1]
```diff
@@ -142,6 +144,22 @@ void MotorHardware::readInputs() {
...
+                case MotorMessage::REG_BATTERY_VOLTAGE: {
+                    int32_t data = mm.getData();
+                    sensor_msgs::BatteryState bstate;
+                    bstate.voltage = (float)data * pid_params.battery_voltage_multiplier +
+                                   pid_params.battery_voltage_offset;
+                    bstate.current = std::numeric_limits<float>::quiet_NaN();
+                    bstate.charge = std::numeric_limits<float>::quiet_NaN();
+                    bstate.capacity = std::numeric_limits<float>::quiet_NaN();
+                    bstate.design_capacity = std::numeric_limits<float>::quiet_NaN();
+                    bstate.percentage = std::numeric_limits<float>::quiet_NaN();
```
Comment (rohbotics) [inline src/motor_hardware.cc]: Can we put in a dumb percentage that goes between 0-100% between 22-28 volts ?

**Review by rohbotics — APPROVED** [review]
Good, but see comments

**Comment by jim-v** [comment]
@rohbotics, how is this?

File: src/motor_hardware.cc [hunk 2]
```diff
@@ -142,6 +144,27 @@ void MotorHardware::readInputs() {
...
+                    int32_t data = mm.getData();
+                    sensor_msgs::BatteryState bstate;
+                    bstate.voltage = (float)data * pid_params.battery_voltage_multiplier +
+                                   pid_params.battery_voltage_offset;
+                    bstate.current = std::numeric_limits<float>::quiet_NaN();
+                    bstate.charge = std::numeric_limits<float>::quiet_NaN();
+                    bstate.capacity = std::numeric_limits<float>::quiet_NaN();
+                    bstate.design_capacity = std::numeric_limits<float>::quiet_NaN();
+                    if (bstate.voltage < 22.0) { // 18V = 0%,  22V = 12%
+                        bstate.percentage = std::max(0.0, 2.0 * (bstate.voltage - 18.0));
```
Comment (rohbotics) [inline src/motor_hardware.cc]: `bstate.percentage` should be between 0 and 1. `0.02 * (bstate.voltage - 18.0)` should make it work

File: src/motor_hardware.cc [hunk 3]
```diff
@@ -142,6 +144,27 @@ void MotorHardware::readInputs() {
...
+                    }
+                    else { // 22V = 12%, 24V = 42%, 26V = 72%. 27V = 87%
+                        bstate.percentage = std::min(100.0, 12.0 + (bstate.voltage - 22.0) * 15.0);
```
Comment (rohbotics) [inline src/motor_hardware.cc]: Same issue, should be something like `std::min(1.0, 0.12 + (bstate.voltage - 22.0) * 0.15);`

**Review by rohbotics — CHANGES_REQUESTED** [review]

File: src/motor_hardware.cc (hunk 2 above)
Comment (rohbotics) [reply src/motor_hardware.cc]: Also I think our lower voltage (0% level) should be 20V. at 18 you are at 9V per battery which is pretty low.

**Comment by jim-v** [comment]
@rohbotics, good point.  I made it a simple linear function with 20V -> 0 and 28V -> 1

File: src/motor_hardware.cc [hunk 4]
```diff
@@ -142,6 +144,27 @@ void MotorHardware::readInputs() {
...
+                    int32_t data = mm.getData();
+                    sensor_msgs::BatteryState bstate;
+                    bstate.voltage = (float)data * pid_params.battery_voltage_multiplier +
+                                   pid_params.battery_voltage_offset;
+                    bstate.current = std::numeric_limits<float>::quiet_NaN();
+                    bstate.charge = std::numeric_limits<float>::quiet_NaN();
+                    bstate.capacity = std::numeric_limits<float>::quiet_NaN();
+                    bstate.design_capacity = std::numeric_limits<float>::quiet_NaN();
+                    if (bstate.voltage < 20.0) 
+                      bstate.percentage = 0.0;
```
Comment (rohbotics) [inline src/motor_hardware.cc]: Why use an if, when you could just use std::max(0.0, bstate.percentage), it would be cleaner I think.

**Review by rohbotics — CHANGES_REQUESTED** [review]

**Review by rohbotics — APPROVED** [review]
LGTM

**Pull request merged**